### DISCUSSION TITLE
[Azure DevOps] The requested pull request was not found

### DISCUSSION THREAD

**Discussion by prufrock** [apache/devlake#5280] [Q&A] [answered]
I've been testing out the Azure DevOps plugin that's in beta, currently using `v0.17.0-beta6`. The Azure DevOps task keeps failing with the following error(I've removed the identifiers for my org):
```
[pipeline service] [pipeline #3] [task #36] [collectAzuredevopsGitpullrequestcommits] pydevlake.api.APIException: APIException: https://dev.azure.com/[org]/[repo]/_apis/git/repositories/[id]/pullRequests/7947/commits?api-version=7.0: 404 body: b'{"$id":"1","innerException":null,"message":"TF401180: The requested pull request was not found.","typeName":"Microsoft.TeamFoundation.Git.Server.GitPullRequestNotFoundException, Microsoft.TeamFoundation.Git.Server","typeKey":"GitPullRequestNotFoundException","errorCode":0,"eventId":3000}'"
[pipeline service] [pipeline #3] [task #36] subtask collectAzuredevopsGitpullrequestcommits ended unexpectedly
Wraps: (2) remote error response:
| ERROR: APIException: APIException: https://dev.azure.com/[org]/[repo]/_apis/git/repositories/[id]/pullRequests/7947/commits?api-version=7.0: 404 body: b'{"$id":"1","innerException":null,"message":"TF401180: The requested pull request was not found.","typeName":"Microsoft.TeamFoundation.Git.Server.GitPullRequestNotFoundException, Microsoft.TeamFoundation.Git.Server","typeKey":"GitPullRequestNotFoundException","errorCode":0,"eventId":3000}'
| Traceback (most recent call last):
|   File "/app/python/plugins/azuredevops/azuredevops/main.py", line 132, in <module>
|     AzureDevOpsPlugin.start()
```
It looks like the PR does still exist in Azure DevOps, so I'm not exactly sure what the problem is. I didn't post this as a bug yet since I am not sure how to reproduce other than pulling data from my Azure DevOps org 😄

Has anybody else bumped into this or been able to ascertain what in Azure DevOps might be causing the trouble?

**Comment by klesh** [comment]
@CamilleTeruel Please take a look at this.

**Reply by prufrock** [reply]
On second look it appears the pull request does not exist. Sorry for the misunderstanding there.

I was able to duplicate the error from the azure cli tool:
```
az devops invoke --area git --resource pullRequestCommits --organization https://dev.azure.com/[org]/ --route-parameters project=[project] repositoryId=[repo]  pullRequestId=7947 --query-parameters  api-version=7.0
TF401180: The requested pull request was not found.
```

Now what I can't figure out is how Devlake got a hold of a pull request ID that doesn't exist.

**Reply by CamilleTeruel** [reply]
Hi @prufrock,

Can you confirm whether `https://dev.azure.com/[org]/[repo]/_apis/git/repositories/[id]/pullRequests` includes PR with id 7947?
If it is indeed absent, does this SQL return a row? 
`SELECT * FROM _tool_azuredevops_gitpullrequests WHERE pull_request_id = 7947`

**Comment by prufrock** [comment]
Hi!

Thanks for looking into my issue.

I don't see PR with id 7947 in the list of PRs at the URL `https://dev.azure.com/[org]/[repo]/_apis/git/repositories/[id]/pullRequests`. 

The SQL query doesn't return any results, but that's likely because I've been attempting to recreate the issue and have cleared out the database a few times. After doing that Devlake is now getting tripped up on a build:

```
ERROR: APIException: APIException: https://dev.azure.com/[org]/[repo]/_apis/build/builds/162748/timeline?api-version=7.0: 204 body: no body"
[pipeline service] [pipeline #10] [task #70] [collectAzuredevopsJobs] Traceback (most recent call last):"
level=error msg=" [pipeline service] [pipeline #10] [task #70] [collectAzuredevopsJobs]   File "/app/python/plugins/azuredevops/azuredevops/main.py", line 132, in <module>"
```

I took a look at this particular build. It looks like the problem is the build encountered an error before it had a chance to run any of the Jobs or Tasks in the pipeline. As a result there wasn't any data to record in the timeline for build 162748.

**Reply by CamilleTeruel** [reply]
Hi prufrock, this 204 error has been fixed by #5300, it should be merged soon.

**Comment by prufrock** [comment]
I was able to recreate it with a different PR and may have found a useful bit clue as to what's wrong.

I had been trying to recreate the issue by isolating the problem down to a single project with a single repo. I had an idea that the problem might be caused by a different repo so I went ahead and added that repo to the list to be imported. After I did that I was able to recreate the issue, albeit with a different PR:

```
[pipeline service] [pipeline #15] [task #110] [collectAzuredevopsGitpullrequestcommits]     raise APIException(response)"
[pipeline service] [pipeline #15] [task #110] [collectAzuredevopsGitpullrequestcommits] pydevlake.api.APIException: APIException: https://dev.azure.com/[org]/[project]/_apis/git/repositories/b200ac05-701f-4e8a-9819-219b532d65d3/pullRequests/10461/commits?api-version=7.0: 404 body: b'{"$id":"1","innerException":null,"message":"TF401180: The requested pull request was not found.","typeName":"Microsoft.TeamFoundation.Git.Server.GitPullRequestNotFoundException, Microsoft.TeamFoundation.Git.Server","typeKey":"GitPullRequestNotFoundException","errorCode":0,"eventId":3000}'"
[pipeline service] [pipeline #15] [task #110] subtask collectAzuredevopsGitpullrequestcommits ended unexpectedly
Wraps: (2) remote error response:
023-05-26 14:26:59       | ERROR: APIException: APIException: https://dev.azure.com/[org]/[project]/_apis/git/repositories/b200ac05-701f-4e8a-9819-219b532d65d3/pullRequests/10461/commits?api-version=7.0: 404 body: b'{"$id":"1","innerException":null,"message":"TF401180: The requested pull request was not found.","typeName":"Microsoft.TeamFoundation.Git.Server.GitPullRequestNotFoundException, Microsoft.TeamFoundation.Git.Server","typeKey":"GitPullRequestNotFoundException","errorCode":0,"eventId":3000}'
```

So this is saying the PR doesn't exist in repo `b200ac05-701f-4e8a-9819-219b532d65d3`, but it does exist just in a different repo. It exists in repo `7cbe19cd-1393-4551-aeb2-becb3160cac5`. Interestingly the `7cbe19cd` id is the same one that shows up in the `scope_id` attribute in `_raw_data_params` and in `url` when I run the query `SELECT * FROM _tool_azuredevops_gitpullrequests WHERE pull_request_id = 10461;`.

Here's what the database has for those two values(`SELECT _raw_data_params, url FROM _tool_azuredevops_gitpullrequests WHERE pull_request_id = 10461;`):
```
"{""connection_id"": 1, ""scope_id"": ""7cbe19cd-1393-4551-aeb2-becb3160cac5""}",https://dev.azure.com/[org]/bf664609-7fc4-4a58-a070-11532d690d14/_apis/git/repositories/7cbe19cd-1393-4551-aeb2-becb3160cac5/pullRequests/10461
```

I hope this helps track down the problem. I'm really excited to get this connected to my organizations Azure DevOps, so let me know if you need more info!

**Reply by CamilleTeruel** [reply]
Thanks this is really helpful info! I'll have a look at this issue next week.

**Reply by prufrock** [reply, marked as answer]
As of v.0.17.0-beta13 this problem appears to be resolved. I've so far been able to import my repos from Azure DevOps without errors. Thanks! :heart:
